src/examples/generation.ts

```diff
@@ -55,40 +59,74 @@ const bone = Armature.define((root: Node) => {
             .rotate(Math.random() * 360)
             .release();
         node
             .hold(node.point('handle'))
-            .rotate(Math.random() * 45)
+            .rotate(Math.random() * 70)
             .release();
         node.scale(0.8); // Shrink a bit
 
-        const trunk = node.point('mid').attach(branchShape);
+        const trunk = instance.add(node.point('mid').attach(branchShape));
         trunk.scale({ x: 0.2, y: 1, z: 0.2 });
 
-        treeGen.addDetail({ component: 'branchOrLeaf', at: node.point('tip') });
+        instance.addDetail({ component: 'branchOrLeaf', at: node.point('tip') });
     })
-    .defineWeighted('branchOrLeaf', 1, (root: Point) => {
-        treeGen.addDetail({ component: 'leaf', at: root });
+    .defineWeighted('branchOrLeaf', 1, (root: Point, instance: GeneratorInstance) => {
+        instance.addDetail({ component: 'leaf', at: root });
     })
-    .defineWeighted('branchOrLeaf', 4, (root: Point) => {
-        treeGen.addDetail({ component: 'branch', at: root });
-        treeGen.addDetail({ component: 'maybeBranch', at: root });
-        treeGen.addDetail({ component: 'maybeBranch', at: root });
+    .defineWeighted('branchOrLeaf', 4, (root: Point, instance: GeneratorInstance) => {
+        instance.addDetail({ component: 'branch', at: root });
+        instance.addDetail({ component: 'maybeBranch', at: root });
+        instance.addDetail({ component: 'maybeBranch', at: root });
     })
-    .define('leaf', (root: Point) => {
-        const leaf = root.attach(leafSphere);
+    .define('leaf', (root: Point, instance: GeneratorInstance) => {
+        const leaf = instance.add(root.attach(leafSphere));
         leaf.scale(Math.random() * 0.5 + 0.5);
     })
-    .maybe('maybeBranch', (root: Point) => {
-        treeGen.addDetail({ component: 'branch', at: root });
+    .maybe('maybeBranch', (root: Point, instance: GeneratorInstance) => {
+        instance.addDetail({ component: 'branch', at: root });
     });
-const tree = treeGen.generate({ start: 'branch', depth: 25 });
+
+/*const tree = treeGen.generateSOSMC({
+    start: 'branch',
+    depth: 150,
+    samples: 100,
+    costFn: CostFunction.forces([
+        {point: {x: -50, y: 100, z: 0}, influence: -200},
+        {point: {x: 0, y: -100, z: 0}, influence: 100}
+    ])
+});*/
+
+const treeTarget = Model.create();
+const sphere = treeTarget.add(new GeometryNode(leafSphere));
+sphere.moveTo({ x: 0, y: 3, z: 0 });
+const branch = treeTarget.add(new GeometryNode(branchShape));
+branch.scale({ x: 0.2, y: 2, z: 0.2 });
+branch.moveTo({ x: 0, y: 1, z: 0 });
+
+const tree = treeGen.generateSOSMC({
+    start: 'branch',
+    depth: 200,
+    samples: 100,
+    costFn: CostFunction.fillVolume(treeTarget, 0.2),
+    onLastGeneration: (instances: GeneratorInstance[]) => {
+        const result = document.createElement('p');
+        result.innerText = 'Costs in final generation: ';
+        result.innerText += instances
```

Comment:
This is complex for the client code we want our users (artists) to write. Is there a better way to do this? Could we potentially (in another PR) create helpers to do this and abstract away this complexity?

Reply:
I mean, it is just updating the inner text, which we could do iteratively too. Nevermind, I'm just being dense.

Reply:
This is entirely meant for debugging, I wouldnt worry about it

Reply:
Like a hook so that we can measure the final cost to compare methods with. Users won't really need this ever